mcf-browser/src/TimelineExplorer.tsx

```diff
@@ -0,0 +1,390 @@
+
+  /** A list of possible locations to select from */
+  locationOptions: SelectOption[]
+
+  /** The key for the select component to tell it
+  * to re-render when the locations are loaded in
+  */
+  selectKey: string;
+
+  /** A mapping from a location dcid to its name */
+  locationMapping: LocationMapping;
+
+  /** The list of groups to plot, grouped by stat var */
+  statVarGroups: Series[][];
+
+  /** Tracks which page the user is currently viewing (0-indexed) */
+  page: number;
+}
+
+/** Component to display the timeline explorer */
+class TimelineExplorer extends Component<
+  TimelineExplorerPropType,
+  TimelineExplorerStateType
+> {
+  /** Constructor for class, sets initial state
+    * @param {TimelineExplorerPropType} props the props passed in by parent
+    * component
+    */
+  constructor(props: TimelineExplorerPropType) {
+    super(props);
+    this.state = {
+      locations: [],
+      locationOptions: [],
+      selectKey: 'unloaded',
+      locationMapping: {},
+      statVarGroups: [],
+      page: 0,
+    };
+  }
+
+  /**
+   * Update the state variables whenever necessary given
+   * the current value of this.props
+   */
+  updateState() {
+    const locationsPromise = this.getAllLocations();
+    locationsPromise.catch(
+        () => {
+          console.log(
+              'Error getting locations for data:',
+          );
+          console.log(this.props.data);
+        },
+    );
+    locationsPromise.then(
+        (locationOptions) => {
+          const locations = locationOptions.slice(0, INITIAL_NUM_LOCATIONS)
+              .map(
+                  (option) => option.value,
+              );
+          const selectKey = locations.join(',');
+          const locationMapping: LocationMapping = {};
+          locationOptions.map(
+              (obj) => locationMapping[obj.value] = obj.label,
+          );
+          const page = 0;
+          const statVarGroups = Object.values(
+              this.getVariableToSeries(locations),
+          ) as Series[][];
+          this.setState(() => {
+            return {
+              locations,
+              locationOptions,
+              selectKey,
+              locationMapping,
+              statVarGroups,
+              page,
+            };
+          });
+        },
+    );
+  }
+
+  /** Tracks when component is mounted */
+  componentDidMount() {
+    this.updateState();
+  }
+
+  /** Set the location options
+    * @param {TimelineExplorerPropType} prevProps the previous props
+   */
+  componentDidUpdate(prevProps: TimelineExplorerPropType) {
+    if (prevProps.data !== this.props.data) {
+      this.updateState();
+    }
+  }
+
+  /** Returns the series objects which match the current locations filter
+    * @param {string[]} locations a list of locations to filter by
+    * @return {Series[]} series that match the filter
+   */
+  getFilteredData(locations: string[]): Series[] {
+    const cleanedLocations = new Set(locations);
```

Comment:
nit: this wouldn't be cleanedLocations right? this is jsut locations but in set form, maybe `locationSet`

Reply:
Changed!